docs: design doc for bucket policy feature (#5800)
rook/rook: design/ceph/object/bucketpolicy.md

target-version: release-1.4
---

# Feature Name
Adding bucket policy support for ceph object store

## Summary
The bucket policy is the feature in which permissions for specific user can be set on s3 bucket. Read more about it from [ceph documentation](https://docs.ceph.com/docs/master/radosgw/bucketpolicy/)

Currently [ceph object store](/Documentation/ceph-object.md) can be consumed either via [OBC](/Documentation/ceph-object-bucket-claim.md) and [ceph object user](/Documentation/ceph-object-store-user-crd.md). As of now there is no direct way for ceph object user to access the OBC. The idea behind this feature to allow that functionality via Rook. Refer bucket policy examples from [here](https://docs.aws.amazon.com/AmazonS3/latest/dev/example-bucket-policies.html).  Please note it is different from `IAM policies`.

## Proposal details

Followings CRDs are needed to add for defining policies:
  - _bucketPolicy_ in the `Spec` section of `CephObjectStoreUser`
  - _bucketPolicy_ in the `parameters` section of `StorageClass` for `ObjectBucketClaim`

Policies need to provide in generic `json` format. A policy can have multiple `statement`.

> Following checks are need to perform in Rook to verify whether the `bucketpolicy` applicable to OBC.

Reviewer: ```suggestion
The following checks are needed to perform in Rook to verify whether the `bucketpolicy` applicable to OBC.
```

Author: Done

> - for ceph object user, `Principal` value should have username and `Resource` should have specific bucket names. It can be defined for buckets which are not part of the OBC as well, the `bucketname` of a OBC can be fetch from its `configmap`.

Reviewer: ```suggestion
- for ceph object user, `Principal` value should have a username and `Resource` should have specific bucket names. It can be defined for buckets which are not part of the OBC as well, the `bucketname` of an OBC can be fetched from its `configmap`.
```

Author: Done

- In `StorageClass`, `Principal` value should be `*`(applicable to all users) and `Resource` should have the bucket name can be empty since it can be generated name from Rook as well and will be attached before setting the policy.

> Examples:-

Reviewer: ```suggestion
Examples:
```

Author: Done

> ```console

Reviewer: ```suggestion
```yaml
```

Author: Done

> # cat storageclass-bucket.yaml

Reviewer: you can remove the `cat`.

Author: Done

apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
   name: rook-ceph-delete-bucket
provisioner: rook-ceph.ceph.rook.io/bucket
# set the reclaim policy to delete the bucket and all objects
# when its OBC is deleted.
reclaimPolicy: Delete
parameters:
  objectStoreName: my-store
  objectStoreNamespace: rook-ceph
  region: us-east-1
  bucketName: ceph-bkt
  bucketPolicy: "Version": "2012-10-17",
		"Statement": [{
		"Effect": "Allow",
		"Principal": {"AWS": ["arn:aws:iam:::*"]},
		"Action": "s3:PutObjectAcl",
		"Resource": [
		"arn:aws:s3:::/*"
		]
	}]

> # cat object-user.yaml

Reviewer: remove the `cat`

Author: Done

---
title: adding bucket policy for ceph object store
target-version: release-1.4
---

# Feature Name
Adding bucket policy support for ceph object store

## Summary
The bucket policy is the feature in which permissions for specific user can be set on s3 bucket. Read more about it from [ceph documentation](https://docs.ceph.com/docs/master/radosgw/bucketpolicy/)

> Currently [ceph object store](/Documentation/ceph-object.md) can be consumed either via [OBC](/Documentation/ceph-object-bucket-claim.md) and [ceph object user](/Documentation/ceph-object-store-user-crd.md). As of now there is no direct way for ceph object user to access the OBC. The idea behind this feature to allow that functionality via Rook. Refer bucket policy examples from [here](https://docs.aws.amazon.com/AmazonS3/latest/dev/example-bucket-policies.html).  Please note it is different from `IAM policies`.

Reviewer: Can we link some doc about those `IAM policies`?

Author: Sure will address in next PR

---
title: adding bucket policy for ceph object store
target-version: release-1.4
---

# Feature Name
Adding bucket policy support for ceph object store

## Summary
The bucket policy is the feature in which permissions for specific user can be set on s3 bucket. Read more about it from [ceph documentation](https://docs.ceph.com/docs/master/radosgw/bucketpolicy/)

Currently [ceph object store](/Documentation/ceph-object.md) can be consumed either via [OBC](/Documentation/ceph-object-bucket-claim.md) and [ceph object user](/Documentation/ceph-object-store-user-crd.md). As of now there is no direct way for ceph object user to access the OBC. The idea behind this feature to allow that functionality via Rook. Refer bucket policy examples from [here](https://docs.aws.amazon.com/AmazonS3/latest/dev/example-bucket-policies.html).  Please note it is different from `IAM policies`.

## Proposal details

> Followings CRDs are needed to add for defining policies:

Reviewer 2: ```suggestion
The following settings are needed to add for defining policies:
```

>   - _bucketPolicy_ in the `Spec` section of `CephObjectStoreUser`

Reviewer 2: ```suggestion
  - _bucketPolicy_ in the `Spec` section of `CephObjectStoreUser` CR
```

Policies need to provide in generic `json` format. A policy can have multiple `statement`.

The following checks are needed to perform in Rook to verify whether the `bucketpolicy` applicable to OBC.
- for ceph object user, `Principal` value should have username and `Resource` should have specific bucket names. It can be defined for buckets which are not part of the OBC as well, the `bucketname` of an OBC can be fetched from its `configmap`.
- In `StorageClass`, `Principal` value should be `*`(applicable to all users) and `Resource` should have the bucket name can be empty since it can be generated name from Rook as well and will be attached before setting the policy.

Examples:
```yaml
# storageclass-bucket.yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
   name: rook-ceph-delete-bucket
provisioner: rook-ceph.ceph.rook.io/bucket
# set the reclaim policy to delete the bucket and all objects
# when its OBC is deleted.
reclaimPolicy: Delete
parameters:
  objectStoreName: my-store
  objectStoreNamespace: rook-ceph
  region: us-east-1
  bucketName: ceph-bkt
  bucketPolicy: "Version": "2012-10-17",
		"Statement": [{

> 		"Effect": "Allow",

Reviewer 2: nit: can you use spaces instead of tabs to avoid this rendering?

Author: will address in next PR

		"Principal": {"AWS": ["arn:aws:iam:::*"]},
		"Action": "s3:PutObjectAcl",
		"Resource": [
		"arn:aws:s3:::/*"
		]
	}]

# object-user.yaml
apiVersion: ceph.rook.io/v1
kind: CephObjectStoreUser
metadata:
  name: my-user
  namespace: rook-ceph
spec:
  store: my-store
  displayName: "my display name"
  bucketPolicy: "Version": "2012-10-17",
		"Statement": [{
		"Effect": "Allow",
		"Principal": {"AWS": ["arn:aws:iam:::my-user"]},
		"Action": "s3:PutObjectAcl",
		"Resource": [
		"arn:aws:s3:::ceph-bkt/*"
		]
	}]
```

The `bucketPolicy` defined on CRD won't override any existing policies on that bucket, will just append. But this can be easily overwritten with help of S3 client, since does not have much control over there.

## Apis and structural changes

The following field will be add to `ObjectStoreUserSpec` and this need to reflected on the existing api's for `CephObjectStoreUser`

```
type ObjectStoreUserSpec struct {
        Store string `json:"store,omitempty"`
        //The display name for the ceph users
        DisplayName string `json:"displayName,omitempty"`
+       //The list of bucket policy for this user
+       BucketPolicy string `json:"bucketPolicy,omitempty"`
 }
```

> The `bucket policy` feature is consumed by the brownfield use case of `OBC`, so supporting apis and structures already exists in [policy.go](/pkg/operator/ceph/object/policy.go). Still few more api's are need to read the policy from CRD, validate it and then convert it into `bucketpolicy`, so that can be consumed by existing api's

Reviewer 2: The ObjectStoreUser CR isn't used with the OBCs, so when would the buckey policy be applied if specified in the CR? I understand for OBCs it can be applied from the storage class when the bucket is created, but I don't see where the operator can apply the policy from the ObjectStoreUser CR.

Author: @travisn Yeah both are different. `bucketpolicy` will be part of the `objectstoreuser` CR itself. At the bucket level both will be applied accordingly. Consider below example : 
```yaml
apiVersion: ceph.rook.io/v1
kind: CephObjectStoreUser
metadata:
  name: my-user
  namespace: rook-ceph
spec:
  store: my-store
  displayName: "my display name"
  bucketPolicy: "Version": "2012-10-17",
		"Statement": [{
		"Effect": "Allow",
		"Principal": {"AWS": ["arn:aws:iam:::my-user"]},
		"Action": "s3:PutObject",
		"Resource": [
		"arn:aws:s3:::ceph-bkt-1/*"
		],
		"Statement": [{
		"Effect": "Allow",
		"Principal": {"AWS": ["arn:aws:iam:::my-user"]},
		"Action": "s3:ListObject",
		"Resource": [
		"arn:aws:s3:::ceph-bkt-2/*"
		],
		"Statement": [{
		"Effect": "Allow",
		"Principal": {"AWS": ["arn:aws:iam:::my-user"]},
		"Action": "s3:GetObject",
		"Resource": [
		"arn:aws:s3:::ceph-bkt-3/*"
		],
	}]
```
In above there are three bucket policies, which need to apply on three different buckets. After creating the user, the policy from CR will be fetched. And on each bucket `putbucketpolicy` will be performed seperately, so we need to make sure that the existing `bucketpolicy`  is not overwritten(just do append)

Reviewer: It's not clear to me either, can you describe the complete workflow? Are we saying that for a given `CephObjectStoreUser` every bucket that would be created or already exist will see their policy appended?

Author: @leseb : The plan is to set `bucketpolicy` for existing buckets, not for the bucket which is created from this user. The main idea is behind this feature to connect the users to existing buckets and from bucket(or OBC) point of view to provide general access to anonymous users. As of now I didn't consider it for the buckets from  `CephObjectStoreUser`. 

So the flow above cases is that, I have three buckets `ceph-bkt-{1..3}` and I want to give different permission for `my-user`(at this point user may or not exists) on them. After updating the policy in yaml file and when request reaches `reconcileCephUser`, it will divide `bucketpolicy` mentioned in CRD into three different policies. Do `GetBucket` on the bucket one by one, fetch the bucket owner's info, fetch existing policy for each bucket if any and finally append applicable policy  on each bucket calls `setbucketpolicy`.

There is similar flow in [Grant](https://github.com/rook/rook/blob/master/pkg/operator/ceph/object/bucket/provisioner.go#L113) in provisioner.go

Reviewer 2: Thanks for the explanation, if you could add that paragraph about the flow in the design doc.

Author: Done

> 1. `func ReadBucketPolicy(CephObjectStoreUser.Spec/storagev1.StorageClass) string {}`

Reviewer 2: No need to mention method names in the design doc. These could change over time, then the design doc would be outdated.

---
title: adding bucket policy for ceph object store
target-version: release-1.4
---

# Feature Name
Adding bucket policy support for ceph object store

## Summary

> The bucket policy is the feature in which permissions for specific user can be set on s3 bucket. Read more about it from [ceph documentation](https://docs.ceph.com/docs/master/radosgw/bucketpolicy/)

Reviewer 3: don't need 'The' here (so "Bucket policy is the...")--but later, "the bucket policy feature" does need "the" and is correct

Currently [ceph object store](/Documentation/ceph-object.md) can be consumed either via [OBC](/Documentation/ceph-object-bucket-claim.md) and [ceph object user](/Documentation/ceph-object-store-user-crd.md). As of now there is no direct way for ceph object user to access the OBC. The idea behind this feature to allow that functionality via Rook. Refer bucket policy examples from [here](https://docs.aws.amazon.com/AmazonS3/latest/dev/example-bucket-policies.html).  Please note it is different from [IAM policies](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies.html).

## Proposal details

The following settings are needed to add for defining policies:
  - _bucketPolicy_ in the `Spec` section of `CephObjectStoreUser` CR
  - _bucketPolicy_ in the `parameters` section of `StorageClass` for `ObjectBucketClaim`

> Policies need to provide in generic `json` format. A policy can have multiple `statement`.

Reviewer 3: to be provided

Reviewer 3: (and following)

Reviewer 3: since they are multiple, "statements" (or, "statement elements" or something

Author: Done

> The following checks are needed to perform in Rook to verify whether the `bucketpolicy` applicable to OBC.

Reviewer 3: would write, "Rook must perform the following checks..."

Author: Done

  region: us-east-1
  bucketName: ceph-bkt
  bucketPolicy: "Version": "2012-10-17",
                "Statement": [
                {
                "Sid": "listobjs",
                "Effect": "Allow",
                "Principal": {"AWS": ["arn:aws:iam:::*"]},
                "Action": "s3:ListObject",
                "Resource": "arn:aws:s3:::/*"
                }
                ]

# object-user.yaml
apiVersion: ceph.rook.io/v1
kind: CephObjectStoreUser
metadata:
  name: my-user
  namespace: rook-ceph
spec:
  store: my-store
  displayName: "my display name"
  bucketPolicy: "Version": "2012-10-17",
                "Statement": [
                {
                "Sid": "putobjs"
                "Effect": "Allow",
                "Principal": {"AWS": ["arn:aws:iam:::my-user"]},
                "Action": "s3:PutObject",
                "Resource": "arn:aws:s3:::ceph-bkt-1/*"
                },
                {
                "Sid": "getobjs"
                "Effect": "Allow",
                "Principal": {"AWS": ["arn:aws:iam:::my-user"]},
                "Action": "s3:GettObject",
                "Resource": "arn:aws:s3:::ceph-bkt-2/*"
                }
                ]
```

> In the above examples, the `bucket policy` mentioned in the `storage class` will be inherited to all the OBCs created from it. And this policy need to be for the anonymous users(all users in the ceph object store), it will attached to bucket during the OBC creation.

Reviewer: ```suggestion
In the above examples, the `bucket policy` mentioned in the `storage class` will be inherited to all the OBCs created from it. And this policy needs to be for the anonymous users(all users in the ceph object store), it will be attached to the bucket during the OBC creation.
```

Author: Done

> Incase of `ceph object store user` the policy can have multiple statement and each represents policy for the existing buckets in the `ceph object store` for the user `my-user`. During creation of the user, the `bucketPolicy` CRD will convert into and divide into different bucket policy statement, then fetch each bucket info and using the credentials of bucket owner this policy will be set via s3 api.

Reviewer: ```suggestion
In the case of `ceph object store user` the policy can have multiple statements and each represents a policy for the existing buckets in the `ceph object store` for the user `my-user`. During the creation of the user, the `bucketPolicy` CRD will convert into and divide into different bucket policy statements, then fetch each bucket info, and using the credentials of bucket owner this policy will be set via s3 API.
```

Author: Done

> The `bucketPolicy` defined on CRD won't override any existing policies on that bucket, will just append. But this can be easily overwritten with help of S3 client, since does not have much control over there.

Author: Done

## Apis and structural changes

> The following field will be add to `ObjectStoreUserSpec` and this need to reflected on the existing api's for `CephObjectStoreUser`

Author: Done
